File: Grind75/56-LongestPalindromicSubstring/cpp/LongestPalindromicSubstring.cpp
#include "LongestPalindromicSubstring.h"
#include <string>

using namespace std;

string Solution::longestPalindrome(string s) {
  if (s.size() < 2)
    return s;
  int bestStart = 0, bestLength = 1;
  int N = s.size();
  for (int center = 1; center < N; center++) {
    int l, r, length;
    const int c = center;

    // check odd palindromes
    for (l = r = c; l >= 0 && r < N && s[l] == s[r]; l--,r++) {
      length = r - l + 1;
      if (length > bestLength) {
        bestLength = length;
        bestStart = l;
      }
    }

    // check even palindromes
    for (l = c-1, r = c; l >= 0 && r < N && s[l] == s[r]; l--,r++) {
      length = r - l + 1;
      if (length > bestLength) {
        bestLength = length;
        bestStart = l;
      }
    }
  }
  return s.substr(bestStart, bestLength);
}

File: Grind75/56-LongestPalindromicSubstring/python/LongestPalindromicSubstring.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')
from PrettyPrint import *
from Colors import *

class Solution:
  def longestPalindrome(self, s: str) -> str:
    if len(s) == 0:
      return ''
    bestStart, bestEnd, bestLength = 0, 0, 1
    for c in range(len(s)):
      # Odd length palindromes
      l, r = c, c
      while l >= 0 and r < len(s) and s[l] == s[r]:
        length = r - l + 1
        if length > bestLength:
          bestLength = length
          bestStart = l
          bestEnd = r
        l -= 1
        r += 1
      # Even length palindromes
      l, r = c-1, c
      while l >= 0 and r < len(s) and s[l] == s[r]:
        length = r - l + 1
        if length > bestLength:
          bestLength = length
          bestStart = l
          bestEnd = r
        l -= 1
        r += 1
    return s[bestStart:bestEnd+1]

if __name__ == '__main__':
  soln = Solution()
  testCases = [
    ("", ""),
    ("a", "a"),
    ("ab", "a"),
    ("aa", "aa"),
    ("aaa", "aaa"),
    ("aaaa", "aaaa"),
    ("aaaaa", "aaaaa"),
    ("aaaaaa", "aaaaaa"),
    ("aba", "aba"),
    ("aaba", "aba"),
    ("abaa", "aba"),
    ("abba", "abba"),
    ("aabba", "abba"),
    ("abbaa", "abba"),
  ]

  for s, answer in testCases:
    def quoted(s: str): return "'" + s + "'"
    colorPrint(LIGHT_BLUE, quoted(s), "=>", quoted(answer), end=' ')
    result = soln.longestPalindrome(s)
    print('-->', end=' ')
    colorPrint(GREEN if result == answer else RED, quoted(result))

